pkg/storage/gcp/saver.go

```diff
@@ -0,0 +1,75 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/option"
+)
+
+// Storage implements the (github.com/gomods/pkg/storage).Saver interface
+type Storage struct {
+	bucket *storage.BucketHandle
+}
+
+// New returns a new Storage instance
+// authenticated using the provided ClientOptions for the associated bucket
+func New(bucketname string, cred option.ClientOption) (*Storage, error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new client: %s", err)
+	}
+	bkt := client.Bucket(bucketname)
+	return &Storage{bucket: bkt}, nil
+}
+
+// Save uploads the modules .mod, .zip and .info files for a given version
+func (s *Storage) Save(module, version string, mod, zip, info []byte) error {
+	ctx := context.Background()
+	errs := make(chan error, 3)
+	// dispatch go routine for each file to upload
+	go save(ctx, errs, s.bucket, module, version, "mod", mod)
+	go save(ctx, errs, s.bucket, module, version, "zip", zip)
+	go save(ctx, errs, s.bucket, module, version, "info", info)
+
+	errsOut := make([]string, 3)
+	// wait for each routine above to send a value
+	for count := 0; count < 3; count++ {
+		err := <-errs
+		if err != nil {
+			errsOut = append(errsOut, err.Error())
+		}
+	}
+
+	// return concatenated error string if there is anything to report
+	if len(errsOut) < 0 {
```

Comment:
when is len<0?
errsOut is initialized with length set to 3
use `errsOut := make([]string, 0, 3)`
also with make([]string, 3) and append you will end up with len==4 and value ["", "", "", "appended_value"]

Reply:
Nice catch, I'm going over it all again before I write over this branch. Trying to keep a clean commit history.

Reply:
it will be squashed into a single commit, so no need to clean the history just code :D

Reply:
Haha oh 🤦‍♂️

Reply:
>Nice catch, I'm going over it all again before I write over this branch. Trying to keep a clean commit history.

I always think this too ...